api/src/opentrons/protocol_engine/state/_labware_origin_math.py

```diff
@@ -171,6 +216,194 @@ def _get_parent_deck_item_origin_to_child_labware_placement_origin(
+            else:
+                raise NotImplementedError(f"Spring force: {spring_force}")
+        else:
+            return _parent_origin_to_slot_bottom_center(
+                parent_deck_item
+            ) + slot_bottom_center_to_child_origin(child_labware)
+    else:
+        # TODO(jh, 06-25-25): This is a temporary shim to unblock FE usage with LW Def3 and more accurately diff
+        #  ongoing positioning snapshot changes, but we should throw an error  after adding all locating features
+        #  if no appropriate LF pair is found.
+        return Point(0, 0, 0)
+
+
+def _get_spring_force(
+    child_labware: LabwareDefinition3,
+    parent_deck_item: _Labware3SupportedParentDefinition,
+) -> SpringDirectionalForce | None:
+    """Returns whether the parent-child stackup has a spring that affects positioning."""
+    assert parent_deck_item.features.get("slotFootprintAsParent") is not None
+    assert child_labware.features.get("slotFootprintAsChild") is not None
+
+    parent_spring_force = parent_deck_item.features["slotFootprintAsParent"].get(
+        "springDirectionalForce"
+    )
+    child_spring_force = child_labware.features["slotFootprintAsChild"].get(
+        "springDirectionalForce"
+    )
+
+    if parent_spring_force is not None and child_spring_force is not None:
+        if parent_spring_force != child_spring_force:
+            raise ValueError(
+                f"Parent spring force: {parent_spring_force} does not match child spring force: {child_spring_force}"
+            )
+
+    return parent_spring_force or child_spring_force
+
+
+def _parent_origin_to_slot_bottom_center(
+    parent_deck_item: _Labware3SupportedParentDefinition,
+) -> Point:
+    """Returns offset from the parent's origin to the slot's bottom center."""
+    slot_footprint_as_parent = parent_deck_item.features.get("slotFootprintAsParent")
+    assert slot_footprint_as_parent is not None
+
+    x = (
+        slot_footprint_as_parent["frontRight"]["x"]
+        + slot_footprint_as_parent["backLeft"]["x"]
+    ) / 2
+    y = (
+        slot_footprint_as_parent["frontRight"]["y"]
+        + slot_footprint_as_parent["backLeft"]["y"]
+    ) / 2
+    z = slot_footprint_as_parent["z"]
+
+    return Point(x, y, z)
+
+
+def _parent_origin_to_slot_back_left_bottom(
+    parent_deck_item: _Labware3SupportedParentDefinition,
+) -> Point:
+    """Returns offset from the parent origin to slot back left bottom."""
+    slot_footprint_as_parent = parent_deck_item.features.get("slotFootprintAsParent")
+    assert slot_footprint_as_parent is not None
+
+    x = slot_footprint_as_parent["backLeft"]["x"]
+    y = slot_footprint_as_parent["backLeft"]["y"]
+    z = slot_footprint_as_parent["z"]
+
+    return Point(x, y, z)
+
+
+def slot_bottom_center_to_child_origin(
+    child_labware: LabwareDefinition3,
+) -> Point:
+    """Returns offset from the slot's bottom center to the child origin."""
```

Comment:
Minor suggestion for clarity, feel free to discard it if you prefer the original.

```suggestion
    """Returns offset from a parent slot's bottom center to the child origin."""
```